File: python/code/4-convolution.py
import numpy as np
import matplotlib.pyplot as plt

# Input sequences
x = np.array([1, 2, 3, 4])
h = np.array([1, 1, 1])

# Convolution
y = np.convolve(x, h)

# Index
n = np.arange(len(y))

print("x[n] =", x)
print("h[n] =", h)
print("y[n] = x[n] * h[n] =", y)

# Plot
plt.figure(figsize=(10, 6))

plt.subplot(3, 1, 1)
plt.stem(np.arange(len(x)), x)
plt.title("Input Sequence x[n]")
plt.grid(True)

plt.subplot(3, 1, 2)
plt.stem(np.arange(len(h)), h)
plt.title("Impulse Response h[n]")
plt.grid(True)

plt.subplot(3, 1, 3)
plt.stem(n, y)
plt.title("Convolution y[n] = x[n] * h[n]")
plt.xlabel("n")
plt.grid(True)

plt.tight_layout()
plt.show()

File: matlab/code/4-convolution.m
clc;
clear;
close all;

% First sequence
x = [1 2 3 4];

% Second sequence
h = [1 1 1];

% Convolution using MATLAB built-in function
y = conv(x, h);

% Time indices
nx = 0:length(x)-1;
nh = 0:length(h)-1;
ny = 0:length(y)-1;

% Display results
disp('x[n] = ');
disp(x);

disp('h[n] = ');
disp(h);

disp('Convolution y[n] = x[n] * h[n] = ');
disp(y);

% Plot x[n]
figure;

subplot(3,1,1);
stem(nx, x, 'filled');
grid on;
xlabel('n');
ylabel('x[n]');
title('Input Sequence x[n]');

% Plot h[n]
subplot(3,1,2);
stem(nh, h, 'filled');
grid on;
xlabel('n');
ylabel('h[n]');
title('Impulse Response h[n]');

% Plot convolution
subplot(3,1,3);
stem(ny, y, 'filled');
grid on;
xlabel('n');
ylabel('y[n]');
title('Convolution y[n] = x[n] * h[n]');